Opensearch integration (#7232)
aws-amplify/docs: src/pages/[platform]/build-a-backend/data/custom-business-logic/search-and-aggregate-queries/index.mdx

AWS OpenSearch Service provides a managed platform for deploying search and analytics solutions with OpenSearch or Elasticsearch. The zero-ETL integration between Amazon DynamoDB and OpenSearch Service allows seamless search on DynamoDB data by automatically replicating and transforming it without requiring custom code or infrastructure. This integration simplifies processes and reduces the operational workload of managing data pipelines.

DynamoDB users gain access to advanced OpenSearch features like full-text search, fuzzy search, auto-complete, and vector search for machine learning capabilities. Amazon OpenSearch Ingestion synchronizes data between DynamoDB and OpenSearch Service, enabling near-instant updates and comprehensive insights across multiple DynamoDB tables. Developers can adjust index mapping templates to match Amazon DynamoDB fields with OpenSearch Service indexes. 

Amazon OpenSearch Ingestion, combined with S3 exports and DynamoDB streams, facilitates seamless data input from DynamoDB tables and automatic ingestion into OpenSearch. Additionally, the pipeline can back up data to S3 for potential future re-ingestion as needed.

Pre-requisites:

> 1. A Todo API created with required access.

Reviewer: "with required access" - I think this reads a bit odd. It's hard to really grasp what this means. Might as well just be explicit on the schema. and state in this guide, we'll sync a Todo table from DynamoDB to OpenSearch

> Important considerations:

Reviewer: Just make this and the previous comment "step 1". I like the explanation below as to why it's needed.

Ensure Point in Time Recovery (PITR) is enabled, which is crucial for the pipeline integration.
Enable DynamoDB streams to capture item changes that will be ingested into OpenSearch.

```ts title="amplify/backend.ts"

> 

Reviewer: show the backend = ...

const todoTable =
  backend.data.resources.cfnResources.amplifyDynamoDbTables["Todo"];

todoTable.pointInTimeRecoveryEnabled = true;

todoTable.streamSpecification = {
  streamViewType: dynamodb.StreamViewType.NEW_IMAGE,
};

```

## Step 1: Setting Up the OpenSearch Instance

Create an OpenSearch instance with encryption.

``` ts title="amplify/backend.ts"
const openSearchDomain = new opensearch.Domain(
  openSearchStack,
  "OpenSearchDomain",
  {
    version: opensearch.EngineVersion.OPENSEARCH_2_11,
    nodeToNodeEncryption: true,
    encryptionAtRest: {
      enabled: true,
    },
  }
);

```

## Step 2: Setting Up Zero ETL from DynamoDB to OpenSearch

### Step 2a: S3 Bucket and IAM Role
Create an S3 bucket for backing up raw events consumed by the OpenSearch pipeline. Also, set up an IAM role for the pipeline:

``` ts title="amplify/backend.ts"

> 

Reviewer: add imports as well on the top

const s3BackupBucket = new s3.Bucket(
  openSearchStack,
  "OpenSearchBackupBucketAmplifyGen2",
  {
    blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
    bucketName: "opensearch-backup-bucket-amplify-gen-2-test",
    enforceSSL: true,
    versioned: true,
    autoDeleteObjects: true,
    removalPolicy: RemovalPolicy.DESTROY,
  }
);

const openSearchIntegrationPipelineRole = new iam.Role(
  openSearchStack,
  "OpenSearchIntegrationPipelineRole",
  {
    assumedBy: new iam.ServicePrincipal("osis-pipelines.amazonaws.com"),
    inlinePolicies: {
      openSearchPipelinePolicy: new iam.PolicyDocument({
        statements: [
          new iam.PolicyStatement({
            actions: ["es:DescribeDomain"],
            resources: [openSearchDomain.domainArn],
            effect: iam.Effect.ALLOW,
          }),
          new iam.PolicyStatement({
            actions: ["es:ESHttp*"],
            resources: [openSearchDomain.domainArn],
            effect: iam.Effect.ALLOW,
          }),
        ],
      }),
    },
    managedPolicies: [
      iam.ManagedPolicy.fromAwsManagedPolicyName(

>         "AmazonOpenSearchServiceFullAccess"

Reviewer: can we scope this down any further?

      ),
    ],
  }
);

openSearchIntegrationPipelineRole.addManagedPolicy(
  iam.ManagedPolicy.fromAwsManagedPolicyName("AmazonDynamoDBFullAccess")
);
openSearchIntegrationPipelineRole.addManagedPolicy(
  iam.ManagedPolicy.fromAwsManagedPolicyName("AmazonS3FullAccess")

> );

Reviewer: same question as above

```
For the S3 bucket, follow standard security practices: block public access, encrypt data at rest, and enable versioning.

> The IAM role should allow the OpenSearch Ingestion Service (OSIS) pipelines to assume it. Grant specific OpenSearch Service permissions and also provide DynamoDB and S3 access. You may customize permissions to follow the principle of least privilege.

Reviewer: "You may customize permissions to follow the principle of least privilege"

let's have the sample show this

### Step 2b: OpenSearch Service Pipeline

Define the pipeline construct and its configuration.

The configuration is a data-prepper feature of OpenSearch. For specific documentation on DynamoDB configuration, refer to [OpenSearch documentation](https://opensearch.org/docs/latest/data-prepper/pipelines/configuration/sources/dynamo-db/).

``` ts title="amplify/backend.ts"

const openSearchTemplate = `
version: "2"
dynamodb-pipeline:
  source:
    dynamodb:
      acknowledgments: true
      tables:
        - table_arn: "${backend.data.resources.tables["Todo"].tableArn}"
          # Remove the stream block if only export is needed
          stream:
            start_position: "LATEST"
          # Remove the export block if only stream is needed
          export:
            s3_bucket: "${s3BackupBucket.bucketName}"

>             s3_region: "us-east-2"

Reviewer: this should probably also be parameterized

>             s3_prefix: "<tableName>/"

Reviewer: same as this?

      aws:
        sts_role_arn: "${openSearchIntegrationPipelineRole.roleArn}"

>         region: "us-east-2"

Reviewer: same here

  sink:
    - opensearch:
        hosts:
          [
            "https://${openSearchDomain.domainEndpoint}",
          ]
        index: "${indexName}"

>         index_type: "custom"

Reviewer: this works well for indexing a single table, what's the best practice for indexing multiple tables? Do you create separate pipelines? Different indexes?

Author: added reference link for indexing multiple tables

For the source configuration, you specify DynamoDB as the data source, identifying the table you want to ingest from and the starting point of the stream. Besides ingesting the stream into OpenSearch, you also define the target S3 bucket for backup purposes. Additionally, you set the IAM role for the ingestion pipeline, making sure it has the required permissions and policies as outlined in the documentation.

For the sink configuration, you specify the OpenSearch domain cluster by setting the host, the index name and type, and template_content (the index mapping) for data formatting. You also configure document-related metadata and set the maximum bulk size for requests to OpenSearch in MB. Again, you specify the IAM role for the sink portion of the pipeline.

When using OpenSearch, you can define the index template or mapping in advance based on your data structure, which allows you to set data types for each field in the document. This approach can be incredibly powerful for precise data ingestion and search. For more information on index mapping/templates, please refer to [OpenSearch documentation](https://opensearch.org/docs/latest/im-plugin/index-templates/).

> Customize the `template_content` JSON-representation to define the data structure for the ingestion pipeline.

Reviewer: sequencing wise, since this is needed before the previous code snippet, shouldn't we show talk about the template_content before?

Author: moved this section prior to configuration

        logGroup: logGroup.logGroupName,
      },
    },
  }
);

```

After deploying the resources, you can test the data ingestion process by adding an item to the `Todo` table. However, before doing that, let's verify that the pipeline has been set up correctly.

In the AWS console, navigate to OpenSearch and then to the pipelines section. You should find your configured pipeline and review its settings to ensure they match your expectations:

![A screenshot of OpenSearch OSIS pipeline](/images/gen2/opensearch-integration/OpenSearch_pipeline.png)

You can also check this in the DynamoDB console by going to the Integrations section of the tables.

![A screenshot of DynamoDB integration](/images/gen2/opensearch-integration/OpenSearch_DynamoDB_integration.png)

## Step 3: Expose new queries on OpenSearch

### Step 3a: Add the AppSync Resolver for the Search Query

First, update the schema and add a searchTodo query.

``` ts title="amplify/data/resource.ts"

const schema = a.schema({
  Todo: a
    .model({
      content: a.string(),
      done: a.boolean(),
      priority: a.enum(["low", "medium", "high"]),
    })
    .authorization([a.allow.public()]),
  //highlight-start  

>   searchTodos: a.query().returns(a.ref("Todo")),

Reviewer: shouldn't this be an array?

  //highlight-end
});
```

> ### Step3b:Add OpenSearch Datasource to backend

Reviewer: ```suggestion
### Step 3b:Add OpenSearch Datasource to backend
```

const osDataSource = backend.data.addOpenSearchDataSource(
  "osDataSource",
  openSearchDomain
);
```
### Step 3c: Create Resolver and attach to query

Create the search resolver and attach it to the searchTodo query. Also, grant the necessary access permissions to OpenSearch.

```ts title="amplify/backend.ts"

new appsync.CfnResolver(openSearchStack, "searchBlogResolver", {
  typeName: "Query",
  fieldName: "searchTodos",
  dataSourceName: "osDataSource",
  apiId: "<app id>",
  runtime: {
    name: "APPSYNC_JS",
    runtimeVersion: "1.0.0",
  },
  code: `import { util } from '@aws-appsync/utils'
  /**
   * Searches for documents by using an input term
   * @param {import('@aws-appsync/utils').Context} ctx the context
   * @returns {*} the request
   */
  export function request(ctx) {
    return {
      operation: 'GET',
      path: "/todo/_search",
    }
  }

  /**
   * Returns the fetched items
   * @param {import('@aws-appsync/utils').Context} ctx the context
   * @returns {*} the result
   */
  export function response(ctx) {
    if (ctx.error) {
      util.error(ctx.error.message, ctx.error.type)
    }
    return ctx.result.hits.hits.map((hit) => hit._source)
  }
  `,

> });

Reviewer: don't do this. instead use `a.handler.custom({ ... })`

import { getCustomStaticPath } from '@/utils/getCustomStaticPath';

export const meta = {
  title: 'Build search and aggregate queries',
  description:

>     'Build search and aggregate queries.',

Reviewer 2: We can extend the description:
```suggestion
    'Build search and aggregate queries with Amazon OpenSearch Service.',
```

  platforms: [
    'android',
    'angular',
    'flutter',
    'javascript',
    'nextjs',
    'react',
    'react-native',
    'swift',
    'vue'
  ]
};

export const getStaticPaths = async () => {
  return getCustomStaticPath(meta.platforms);
};

export function getStaticProps() {
  return {
    props: {
      meta
    }
  };
}

# OpenSearch Integration

> AWS OpenSearch Service provides a managed platform for deploying search and analytics solutions with OpenSearch or Elasticsearch. The zero-ETL integration between Amazon DynamoDB and OpenSearch Service allows seamless search on DynamoDB data by automatically replicating and transforming it without requiring custom code or infrastructure. This integration simplifies processes and reduces the operational workload of managing data pipelines.

Reviewer 2: ```suggestion
Amazon OpenSearch Service provides a managed platform for deploying search and analytics solutions with OpenSearch or Elasticsearch. The zero-ETL integration between Amazon DynamoDB and OpenSearch Service allows seamless search on DynamoDB data by automatically replicating and transforming it without requiring custom code or infrastructure. This integration simplifies processes and reduces the operational workload of managing data pipelines.
```

DynamoDB users gain access to advanced OpenSearch features like full-text search, fuzzy search, auto-complete, and vector search for machine learning capabilities. Amazon OpenSearch Ingestion synchronizes data between DynamoDB and OpenSearch Service, enabling near-instant updates and comprehensive insights across multiple DynamoDB tables. Developers can adjust index mapping templates to match Amazon DynamoDB fields with OpenSearch Service indexes. 

Amazon OpenSearch Ingestion, combined with S3 exports and DynamoDB streams, facilitates seamless data input from DynamoDB tables and automatic ingestion into OpenSearch. Additionally, the pipeline can back up data to S3 for potential future re-ingestion as needed.

## Step 1: Setup the project

> Begin by setting up your project by following the instructions in the [Quickstart guide](https://github.com/aws-amplify/docs/blob/dc76911faee093d7ceaf8a5bf61b54d5f49e42fb/%5Bplatform%5D/start/quickstart). For the purpose of this guide, we'll sync a Todo table from DynamoDB to OpenSearch.

Reviewer 2: Why is there a githun link here?

Author: updated the correct URL

Firstly, add the Todo model to your schema:

```ts title="amplify/data/resource.ts"

import { type ClientSchema, a, defineData } from "@aws-amplify/backend";

const schema = a.schema({
  // highlight-start
  Todo: a
    .model({
      content: a.string(),
      done: a.boolean(),
      priority: a.enum(["low", "medium", "high"]),
    })

>     .authorization([a.allow.public()]),

Reviewer: new authorization rule "allow => [allow.publicApiKey]"

Author: updated

  // highlight-end  
});

export type Schema = ClientSchema<typeof schema>;

export const data = defineData({
  schema,
  authorizationModes: {
    defaultAuthorizationMode: "apiKey",
    apiKeyAuthorizationMode: {
      expiresInDays: 30,
    },
  },
});

```

Important considerations:

Ensure Point in Time Recovery (PITR) is enabled, which is crucial for the pipeline integration.
Enable DynamoDB streams to capture item changes that will be ingested into OpenSearch.

```ts title="amplify/backend.ts"

import { defineBackend } from "@aws-amplify/backend";
import { auth } from "./auth/resource";
import { data } from "./data/resource";
import * as dynamodb from "aws-cdk-lib/aws-dynamodb";

const backend = defineBackend({
  auth,
  data,
});

const todoTable =
  backend.data.resources.cfnResources.amplifyDynamoDbTables["Todo"];

todoTable.pointInTimeRecoveryEnabled = true;

todoTable.streamSpecification = {
  streamViewType: dynamodb.StreamViewType.NEW_IMAGE,
};

```

## Step 2: Setting Up the OpenSearch Instance

Create an OpenSearch instance with encryption.

``` ts title="amplify/backend.ts"
const openSearchDomain = new opensearch.Domain(

>   openSearchStack,

Reviewer: you use this "openSearchStack" here but it's not defined prior.

Author: updated

``` ts title="amplify/backend.ts"
import { defineBackend } from "@aws-amplify/backend";
import { auth } from "./auth/resource";
import { data } from "./data/resource";
import * as dynamodb from "aws-cdk-lib/aws-dynamodb";
import * as opensearch from "aws-cdk-lib/aws-opensearchservice";
import * as osis from "aws-cdk-lib/aws-osis";
import * as s3 from "aws-cdk-lib/aws-s3";
import * as iam from "aws-cdk-lib/aws-iam";
import * as logs from "aws-cdk-lib/aws-logs";
import { RemovalPolicy } from "aws-cdk-lib";
import { PolicyStatement } from "aws-cdk-lib/aws-iam";

const s3BackupBucket = new s3.Bucket(
  openSearchStack,
  "OpenSearchBackupBucketAmplifyGen2",
  {
    blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
    bucketName: "opensearch-backup-bucket-amplify-gen-2-test",
    enforceSSL: true,
    versioned: true,
    autoDeleteObjects: true,
    removalPolicy: RemovalPolicy.DESTROY,
  }
);

> 

Reviewer: any chance we can use a subpath of the `defineStorage` bucket? Does this approach work with multi-environments? I'm not sure if S3 allows the same bucket name (I thought not)

Author: In the openSearchTemplate configuration, we're utilizing the 'tableName' variable as the 's3_prefix'. This results in the pipeline creating indexes within a folder named after the table.

const openSearchIntegrationPipelineRole = new iam.Role(
  openSearchStack,
  "OpenSearchIntegrationPipelineRole",
  {
    assumedBy: new iam.ServicePrincipal("osis-pipelines.amazonaws.com"),
    inlinePolicies: {
      openSearchPipelinePolicy: new iam.PolicyDocument({
        statements: [
          new iam.PolicyStatement({
            actions: ["es:DescribeDomain"],

>             resources: ["arn:aws:es:*:{your-account-id}:domain/*"],

Reviewer: can you add this value as a reference?

Author: updated

            effect: iam.Effect.ALLOW,
          }),
          new iam.PolicyStatement({

>             actions: ["es:ESHttp*"],

Reviewer 2: Should we allow all actions?

>             resources: ["*"],

Reviewer 2: Should we limit the actions to only open search domain instead of all?

Author: Yes, we require access solely to the OpenSearch domain. The code has been updated accordingly.

>             resources: ["arn:aws:es:*:{your-account-id}:domain/{domain-namedomain}/*"],

Reviewer: can you add this value as a reference?

Author: done

            effect: iam.Effect.ALLOW,
          }),
          new iam.PolicyStatement({
            effect: iam.Effect.ALLOW,
            actions: [
              "s3:GetObject",
              "s3:AbortMultipartUpload",
              "s3:PutObject",
              "s3:PutObjectAcl",
            ],
            resources: [
              "<s3Bucket-arn>",

>               "<s3Bucket-arn>/*"

Reviewer: can you add these values as code references?

Author: updated

            ],
          }),
          new iam.PolicyStatement({
            effect: iam.Effect.ALLOW,
            actions: [
              "dynamodb:DescribeTable",
              "dynamodb:DescribeContinuousBackups",
              "dynamodb:ExportTableToPointInTime",
              "dynamodb:DescribeExport",
              "dynamodb:DescribeStream",
              "dynamodb:GetRecords",

>               "dynamodb:GetShardIterator",

Reviewer 2: Do we really need all these permissions to do the search?

Author: Yes, according to this [documentation on OpenSearch](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/configure-client-ddb.html), these permissions are configured to follow the principle of least privilege.

            ],
            resources: [
              "<dynamodb-table-arn>",

>               "<dynamodb-table-arn>/*",

Author: done

In the source configuration, DynamoDB is specified as the data source, along with the target table for ingestion and the starting point of the stream. Additionally, besides ingesting the stream into OpenSearch, a target S3 bucket is defined for backup purposes. Furthermore, an IAM role is set for the ingestion pipeline, ensuring it possesses the necessary permissions and policies as detailed in the documentation.

Regarding the sink configuration, the OpenSearch domain cluster is specified by setting the host, index name, type, and template content (index mapping) for data formatting. Document-related metadata is configured along with the maximum bulk size for requests to OpenSearch in MB. Once again, an IAM role is specified for the sink portion of the pipeline. For further details on Sink configuration, please refer to the [OpenSearch documentation](https://opensearch.org/docs/latest/data-prepper/pipelines/configuration/sinks/sinks/).

The sink configuration is an array. To create a different index on the same table, you can achieve this by adding a second OpenSearch configuration to the sink array.

> "To index multiple tables, you'll need to configure multiple pipelines in the configuration. For further guidance, please consult the [pipeline section](https://opensearch.org/docs/latest/data-prepper/pipelines/pipelines/) of the OpenSearch documentation."

Reviewer: ```suggestion
To index multiple tables, you'll need to configure multiple pipelines in the configuration. For further guidance, please consult the [pipeline section](https://opensearch.org/docs/latest/data-prepper/pipelines/pipelines/) of the OpenSearch documentation.
```

Now, create the OSIS pipeline resource:

``` ts title="amplify/backend.ts"
const logGroup = new logs.LogGroup(openSearchStack, "LogGroup", {
  logGroupName: "/aws/vended-logs/OpenSearchService/pipelines/1",
  removalPolicy: RemovalPolicy.DESTROY,
});

const cfnPipeline = new osis.CfnPipeline(
  openSearchStack,
  "OpenSearchIntegrationPipeline",
  {
    maxUnits: 4,
    minUnits: 1,
    pipelineConfigurationBody: openSearchTemplate,
    pipelineName: "dynamodb-integration-2",
    logPublishingOptions: {
      isLoggingEnabled: true,
      cloudWatchLogDestination: {
        logGroup: logGroup.logGroupName,
      },
    },
  }
);

```

After deploying the resources, you can test the data ingestion process by adding an item to the `Todo` table. However, before doing that, let's verify that the pipeline has been set up correctly.

In the AWS console, navigate to OpenSearch and then to the pipelines section. You should find your configured pipeline and review its settings to ensure they match your expectations:

> ![A screenshot of OpenSearch OSIS pipeline](/images/gen2/opensearch-integration/OpenSearch_pipeline.png)

Reviewer 3: Could we please get a more descriptive alt text? From the above text it looks like we're referencing what's in the image and so we should make the alt text describe that to help users who are using screen readers

Author: updated alt text.

You can also check this in the DynamoDB console by going to the Integrations section of the tables.

> ![A screenshot of DynamoDB integration](/images/gen2/opensearch-integration/OpenSearch_DynamoDB_integration.png)

Reviewer 3: Could we please get a more descriptive alt text for this that's explaining where we are in the UI?

Author: updated alt text.

  return {
    operation: "GET",
    path: "/todo/_search",
  };
}

/**
 * Returns the fetched items
 * @param {import('@aws-appsync/utils').Context} ctx the context
 * @returns {*} the result
 */

export function response(ctx) {
  if (ctx.error) {
    util.error(ctx.error.message, ctx.error.type);
  }
  return ctx.result.hits.hits.map((hit) => hit._source);
}

```

### Step 4c: Add the AppSync Resolver for the Search Query

Update the schema and add a searchTodo query.

``` ts title="amplify/data/resource.ts"

const schema = a.schema({
  Todo: a
    .model({
      content: a.string(),
      done: a.boolean(),
      priority: a.enum(["low", "medium", "high"]),
    })
    .authorization([a.allow.public()]),

  //highlight-start  
    searchTodos: a
    .query()
    .returns(a.ref("Todo").array())
    .authorization([a.allow.public()])
    .handler(
      a.handler.custom({
        entry: "./searchBlogResolver.js",
        dataSource: "osDataSource",
      })
    ),
  //highlight-end

});
```

Once you've deployed the resources, you can check the AppSync console to confirm the changes. Run the searchTodo query and review the returned results to verify their accuracy.

> ![A screenshot of AppSync console](/images/gen2/opensearch-integration/opensearch_appsync_console.png)

Reviewer 3: Could we get a more descriptive alt text for this? Maybe something like,

`A screenshot of the AppSync console, showing a query for .... with the results of ... on the right`

Something like that, maybe?

Author: updated alt text.

> ![Ascreenshot from the AppSync console displaying a generated query for 'searchTodo' with the results fetched from OpenSearch on the right side.](/images/gen2/opensearch-integration/opensearch_appsync_console.png)

Reviewer 2: ```suggestion
![A screenshot from the AppSync console displaying a generated query for 'searchTodo' with the results fetched from OpenSearch on the right side.](/images/gen2/opensearch-integration/opensearch_appsync_console.png)
```

```ts title="amplify/data/searchBlogResolver.ts"

import { util } from "@aws-appsync/utils";

/**
 * Searches for documents by using an input term
 * @param {import('@aws-appsync/utils').Context} ctx the context
 * @returns {*} the request
 */

export function request(ctx) {
  return {
    operation: "GET",
    path: "/todo/_search",
  };
}

/**
 * Returns the fetched items
 * @param {import('@aws-appsync/utils').Context} ctx the context
 * @returns {*} the result
 */

export function response(ctx) {
  if (ctx.error) {
    util.error(ctx.error.message, ctx.error.type);
  }
  return ctx.result.hits.hits.map((hit) => hit._source);
}

```

### Step 4c: Add the AppSync Resolver for the Search Query

Update the schema and add a searchTodo query.

``` ts title="amplify/data/resource.ts"

const schema = a.schema({
  Todo: a
    .model({
      content: a.string(),
      done: a.boolean(),
      priority: a.enum(["low", "medium", "high"]),
    })
    .authorization([a.allow.public()]),

  //highlight-start  
    searchTodos: a
    .query()
    .returns(a.ref("Todo").array())

>     .authorization([a.allow.public()])

Reviewer: do sweep to make sure the new authZ format is used: "allow => allow.publicApiKey()"

Author: corrected the auth format.

Amazon OpenSearch Ingestion, combined with S3 exports and DynamoDB streams, facilitates seamless data input from DynamoDB tables and automatic ingestion into OpenSearch. Additionally, the pipeline can back up data to S3 for potential future re-ingestion as needed.

## Step 1: Setup the project
Begin by setting up your project by following the instructions in the [Quickstart guide](/[platform]/start/quickstart/). For the purpose of this guide, we'll sync a Todo table from DynamoDB to OpenSearch.

Firstly, add the Todo model to your schema:

```ts title="amplify/data/resource.ts"

import { type ClientSchema, a, defineData } from "@aws-amplify/backend";

const schema = a.schema({
  // highlight-start
  Todo: a
    .model({
      content: a.string(),
      done: a.boolean(),
      priority: a.enum(["low", "medium", "high"]),
    })
    .authorization((allow) => [allow.publicApiKey()])
  // highlight-end  
});

export type Schema = ClientSchema<typeof schema>;

export const data = defineData({
  schema,
  authorizationModes: {
    defaultAuthorizationMode: "apiKey",
    apiKeyAuthorizationMode: {
      expiresInDays: 30,
    },
  },
});

```

Important considerations:

> Ensure Point in Time Recovery (PITR) is enabled, which is crucial for the pipeline integration.

Reviewer 4: looks important, would it be better to do a callout here as well

```ts title="amplify/backend.ts"

import { defineBackend } from "@aws-amplify/backend";
import { auth } from "./auth/resource";
import { data } from "./data/resource";
import * as dynamodb from "aws-cdk-lib/aws-dynamodb";

const backend = defineBackend({
  auth,
  data,
});

// highlight-start
const todoTable =
  backend.data.resources.cfnResources.amplifyDynamoDbTables["Todo"];

// Update table settings
todoTable.pointInTimeRecoveryEnabled = true;

todoTable.streamSpecification = {
  streamViewType: dynamodb.StreamViewType.NEW_IMAGE,
};

// Get the DynamoDB table ARN
const tableArn = backend.data.resources.tables["Todo"].tableArn;
// Get the DynamoDB table name
const tableName = backend.data.resources.tables["Todo"].tableName;
// highlight-end 

```

## Step 2: Setting Up the OpenSearch Instance

Create an OpenSearch instance with encryption.

``` ts title="amplify/backend.ts"

import { defineBackend } from "@aws-amplify/backend";
import { auth } from "./auth/resource";
import { data } from "./data/resource";
import * as dynamodb from "aws-cdk-lib/aws-dynamodb";
import * as opensearch from "aws-cdk-lib/aws-opensearchservice";
import * as osis from "aws-cdk-lib/aws-osis";

> import * as s3 from "aws-cdk-lib/aws-s3";

Reviewer 4: s3 import here might not be needed as we later show defineStorage. but the import might need to be cleaned up as this example doesnt use all of them

Author: s3 import can be removed as we changed it to storage. But rest of the imports are required and used.

Reviewer 5: Do we want to add highlights to the new imports, considering this is the same file as the previous code snippet

import * as iam from "aws-cdk-lib/aws-iam";
import * as logs from "aws-cdk-lib/aws-logs";
import { RemovalPolicy } from "aws-cdk-lib";
import {  Stack } from "aws-cdk-lib";

> 

Reviewer 5: It's unclear to me if the piece of code below should be before or after the code shown in the previous code snippet

Author: updated the steps using the highlights.

// Get the data stack

> const openSearchStack = Stack.of(backend.data);

Reviewer 5: Would it be confusing to call the data stack as `openSearchStack`?

Author: changed to datastack

### Step 3a: Setup Storage and IAM Role
Establish Storage to back up raw events consumed by the OpenSearch pipeline. 
Generate a file named `amplify/storage/resource.ts` and insert the provided content to set up a storage resource. Tailor your storage configurations to regulate access to different paths within your storage bucket.

```ts title="amplify/storage/resource.ts"
import { defineStorage } from "@aws-amplify/backend"

export const storage = defineStorage({
  name: "opensearch-backup-bucket-amplify-gen-2-test",
  access: allow => ({
    'public/*': [
      allow.guest.to(['list', 'write', 'get'])
    ]
  })
})

```
Retrieve the `s3BucketArn` and `s3BucketName` values from the storage resource as demonstrated below. Additionally, set up an IAM role for the pipeline and allocate the roles as specified. For additional details on the necessary IAM roles, please consult the [Setting up roles and users documentation](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/pipeline-security-overview.html#pipeline-security-create).

```ts title="amplify/storage/resource.ts"
import { defineStorage } from "@aws-amplify/backend"

export const storage = defineStorage({

>   name: "opensearch-backup-bucket-amplify-gen-2-test",

Reviewer 4: the bucket name can just be `opensearch-backup-bucket` the name with test might not be ideal

Reviewer 4: also it appears there are two defineStorage example with same example

Author: S3 buckets names are global and more unique are better than general names. Also, removed the duplicate second storage section.

  access: allow => ({
    'public/*': [
      allow.guest.to(['list', 'write', 'get'])
    ]
  })
})

```

> Get the `s3BucketArn` and `s3BucketName` values from storage resource as shown below. Additionally, configure an IAM role for the pipeline and assign the roles as indicated below. For further information on the required IAM roles, please refer to the [Setting up roles and users](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/pipeline-security-overview.html#pipeline-security-create) documentation.

Reviewer 5: This looks like a shorter version of line 166.

Author: removed.

``` ts title="amplify/backend.ts"
import { defineBackend } from "@aws-amplify/backend";
import { auth } from "./auth/resource";
import { data } from "./data/resource";

> import * as dynamodb from "aws-cdk-lib/aws-dynamodb";

Reviewer 4: to align with the other docs examples, the imports show similar as follows

```
import { Policy, PolicyStatement } from "aws-cdk-lib/aws-iam";
```

Author: this change requires code changes in multiple areas and testing take a significant amount of time. so not implementing this change

import * as opensearch from "aws-cdk-lib/aws-opensearchservice";

> import * as osis from "aws-cdk-lib/aws-osis";

Reviewer 4: dont see all the imports being used

Author: All three imports are required and been used in the pipeline config.

For the S3 bucket, follow standard security practices: block public access, encrypt data at rest, and enable versioning.

The IAM role should allow the OpenSearch Ingestion Service (OSIS) pipelines to assume it. Grant specific OpenSearch Service permissions and also provide DynamoDB and S3 access. You may customize permissions to follow the principle of least privilege.

### Step 3b: OpenSearch Service Pipeline

Define the pipeline construct and its configuration.

When using OpenSearch, you can define the index template or mapping in advance based on your data structure, which allows you to set data types for each field in the document. This approach can be incredibly powerful for precise data ingestion and search. For more information on index mapping/templates, please refer to [OpenSearch documentation](https://opensearch.org/docs/latest/im-plugin/index-templates/).

Customize the `template_content` JSON-representation to define the data structure for the ingestion pipeline.

``` ts title="amplify/backend.ts"

// Define OpenSearch index mappings
const indexName = "todo";

const indexMapping = {
  settings: {
    number_of_shards: 1,
    number_of_replicas: 0,
  },
  mappings: {

>     properties: {

Reviewer 5: I don't see `priority`. Was that intentional?

Author: updated

      id: {
        type: "keyword",
      },

>       isDone: {

Reviewer 5: Nit: in the model mentioned above, this is called `done`, I'm not sure if that would be a problem.

Author: `done` is a boolean, hence it's referenced as `isDone`. I can see the expected results.

Regarding the sink configuration, the OpenSearch domain cluster is specified by setting the host, index name, type, and template content (index mapping) for data formatting. Document-related metadata is configured along with the maximum bulk size for requests to OpenSearch in MB. Once again, an IAM role is specified for the sink portion of the pipeline. For further details on Sink configuration, please refer to the [OpenSearch documentation](https://opensearch.org/docs/latest/data-prepper/pipelines/configuration/sinks/sinks/).

The sink configuration is an array. To create a different index on the same table, you can achieve this by adding a second OpenSearch configuration to the sink array.

To index multiple tables, you'll need to configure multiple pipelines in the configuration. For further guidance, please consult the [pipeline section](https://opensearch.org/docs/latest/data-prepper/pipelines/pipelines/) of the OpenSearch documentation.

> NOTE: An OpenSearch Ingestion pipeline supports only one DynamoDB table as its source. For more details on current limitations, Please refer to [Amazon OpenSearch Limitation](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/configure-client-ddb.html#ddb-pipeline-limitations) section.

Reviewer 4: Does this need to be a callout?

Author: wrapped in callout

![A screenshot displaying the OpenSearch OSIS pipeline created under the DynamoDB integrations section](/images/gen2/opensearch-integration/OpenSearch_pipeline.png)

You can also check this in the DynamoDB console by going to the Integrations section of the tables.

![A screenshot displaying the OpenSearch OSIS pipeline created within the 'Ingestion -> Pipelines' section of the OpenSearch Console.](/images/gen2/opensearch-integration/OpenSearch_DynamoDB_integration.png)

## Step 4: Expose new queries on OpenSearch

### Step 4a:Add OpenSearch Datasource to backend

First, Add the OpenSearch data source to the data backend.

``` ts title="amplify/backend.ts"
// Add OpenSearch data source 
const osDataSource = backend.data.addOpenSearchDataSource(
  "osDataSource",
  openSearchDomain
);
```
### Step 4b: Create Resolver and attach to query

Let's create the search resolver. Create a new file named `amplify/data/searchBlogResolver.ts` and paste the following code. For additional details please refer to [Amazon OpenSearch Service Resolvers](https://docs.aws.amazon.com/appsync/latest/devguide/tutorial-elasticsearch-resolvers-js.html)

> ```ts title="amplify/data/searchBlogResolver.ts"

Reviewer 4: is the file extension supposed to be js?

Reviewer 5: Also, should the name be `searchTodoResolver` instead?

import { util } from "@aws-appsync/utils";

/**
 * Searches for documents by using an input term
 * @param {import('@aws-appsync/utils').Context} ctx the context
 * @returns {*} the request
 */

export function request(ctx) {
  return {
    operation: "GET",
    path: "/todo/_search",
  };
}

/**
 * Returns the fetched items
 * @param {import('@aws-appsync/utils').Context} ctx the context
 * @returns {*} the result
 */

export function response(ctx) {
  if (ctx.error) {
    util.error(ctx.error.message, ctx.error.type);
  }
  return ctx.result.hits.hits.map((hit) => hit._source);
}

```

### Step 4c: Add the AppSync Resolver for the Search Query

Update the schema and add a searchTodo query.

``` ts title="amplify/data/resource.ts"

const schema = a.schema({
  Todo: a
    .model({
      content: a.string(),
      done: a.boolean(),
      priority: a.enum(["low", "medium", "high"]),
    })
    .authorization((allow) => [allow.publicApiKey()]),

  //highlight-start  
    searchTodos: a
    .query()
    .returns(a.ref("Todo").array())
    .authorization((allow) => [allow.publicApiKey()])
    .handler(
      a.handler.custom({

>         entry: "./searchBlogResolver.js",

Reviewer 5: Ditto

Author: updated

        dataSource: "osDataSource",
      })
    ),
  //highlight-end

});
```

> Once you've deployed the resources, you can check the AppSync console to confirm the changes. Run the searchTodo query and review the returned results to verify their accuracy.

Reviewer 4: ```suggestion
Once you've deployed the resources, you can verify the changes by checking the AppSync console. Run the 'searchTodo' query and review the results to confirm their accuracy.
```

To index multiple tables, you'll need to configure multiple pipelines in the configuration. For further guidance, please consult the [pipeline section](https://opensearch.org/docs/latest/data-prepper/pipelines/pipelines/) of the OpenSearch documentation.

<Callout>
NOTE: An OpenSearch Ingestion pipeline supports only one DynamoDB table as its source. For more details on current limitations, Please refer to [Amazon OpenSearch Limitation](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/configure-client-ddb.html#ddb-pipeline-limitations) section.
</Callout>
Now, create the OSIS pipeline resource:

``` ts title="amplify/backend.ts"
import { defineBackend } from "@aws-amplify/backend";
import { auth } from "./auth/resource";
import { data } from "./data/resource";
import * as dynamodb from "aws-cdk-lib/aws-dynamodb";
import * as opensearch from "aws-cdk-lib/aws-opensearchservice";
import { Stack } from "aws-cdk-lib"; 
import { storage } from "./storage/resource";
import * as iam from "aws-cdk-lib/aws-iam";
 //highlight-start
import * as logs from "aws-cdk-lib/aws-logs";

> import { RemovalPolicy } from "aws-cdk-lib"; 

Reviewer 4: we will need to add the oasis import here

Author: added the import
